src/rsa/key_generation.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright 2018 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+//! RSA key generation
+
+use super::small_primes::SMALL_PRIMES;
+use core::marker::PhantomData;
+use std::vec::Vec;
+
+use crate::{
+    limb::{self, LimbMask},
+    error, rand,
+    rsa::bigint,
+};
+
+/// Performs the Miller-Rabin primality test
+///
+/// Implementation of the algorithm as it
+/// is described in FIPS 186-4 C.3.1
+fn miller_rabin_test(
+    w: &[u64],
```

Comment:
You have written this whole thing in terms of `Limb` but it would be a lot clearer to write it in terms of `bigint::Nonnegative`, I think.